pkg/stacker/publisher.go

```diff
@@ -65,6 +75,197 @@ func NewPublisher(opts *PublishArgs) *Publisher {
 	}
 }
 
+func basicAuth(username, password string) string {
+	auth := username + ":" + password
+	return base64.StdEncoding.EncodeToString([]byte(auth))
+}
+
+func clientRequest(method, url, username, password string, headers map[string]string, body io.Reader) (*http.Response, error) {
+	req, err := http.NewRequestWithContext(context.TODO(), method, url, body)
+	if err != nil {
+		log.Errorf("unable to create http request err:%s", err)
+		return nil, err
+	}
+
+	// FIXME: handle bearer auth also
+	if username != "" && password != "" {
+		req.Header.Add("Authorization", "Basic "+basicAuth(username, password))
+	}
+
+	if len(headers) > 0 {
+		for k, v := range headers {
+			req.Header.Add(k, v)
+		}
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		log.Errorf("http request failed url:%s", url)
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func fileDigest(path string) (*godigest.Digest, error) {
+	fh, err := os.Open(path)
+	if err != nil {
+		log.Errorf("unable to open file:%s, err:%s", path, err)
+		return nil, err
+	}
+	defer fh.Close()
+
+	dgst, err := godigest.FromReader(fh)
+	if err != nil {
+		log.Errorf("unable get digest for file:%s, err:%s", path, err)
+		return nil, err
+	}
+
+	return &dgst, nil
+}
+
+// publishArtifact to a registry/repo for this subject
+func (p *Publisher) publishArtifact(path, mtype, registry, repo, subject string, skipTLS bool) error {
```

Comment:
Are there 'containers/image' library calls that you could use?

Reply:
nope :(